pyi.Dockerfile

```
@@ -31,18 +49,33 @@ RUN python3 -m pip install -r dev-requirements.txt
 COPY scripts/build.sh scripts/build.sh
 RUN ./scripts/build.sh
 
-COPY --from=bcc-builder /bcc/root/share/bcc/examples/cpp/PyPerf gprofiler/resources/python/pyperf/
+# copy PyPerf and stuff
+RUN cp /bcc/root/share/bcc/examples/cpp/PyPerf gprofiler/resources/python/pyperf/
 # copy licenses and notice file.
-COPY --from=bcc-builder /bcc/bcc/LICENSE.txt gprofiler/resources/python/pyperf/
-COPY --from=bcc-builder /bcc/bcc/licenses gprofiler/resources/python/pyperf/licenses
-COPY --from=bcc-builder /bcc/bcc/NOTICE gprofiler/resources/python/pyperf/
+RUN cp /bcc/bcc/LICENSE.txt gprofiler/resources/python/pyperf/
+RUN cp -r /bcc/bcc/licenses gprofiler/resources/python/pyperf/licenses
+RUN cp /bcc/bcc/NOTICE gprofiler/resources/python/pyperf/
+
+COPY gprofiler gprofiler
+
+# run PyInstaller and make sure no 'gprofiler.*' modules are missing.
+# see https://pyinstaller.readthedocs.io/en/stable/when-things-go-wrong.html
+# from a quick look I didn't see how to tell PyInstaller to exit with an error on this, hence
+# this check in the shell.
+COPY pyi_build.py pyinstaller.spec .
+RUN pyinstaller pyinstaller.spec \
+    && echo \
+    && test -f build/pyinstaller/warn-pyinstaller.txt \
+    && if grep 'gprofiler\.' build/pyinstaller/warn-pyinstaller.txt ; then echo 'PyInstaller failed to pack gProfiler code! See lines above. Make sure to check for SyntaxError as this is often the reason.'; exit 1; fi;
 
-COPY . .
-RUN pyinstaller pyinstaller.spec
+COPY ./scripts/list_needed_libs.sh ./scripts/list_needed_libs.sh
 # staticx packs dynamically linked app with all of their dependencies, it tries to figure out which dynamic libraries are need for its execution
 # in some cases, when the application is lazily loading some DSOs, staticx doesn't handle it.
 # libcgcc_s is such a library, so we pass it manually to staticx.
-RUN staticx -l /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so dist/gprofiler dist/gprofiler
+# additionally, we use list_needed_libs.sh to list the dynamic dependencies of *all* of our resources,
+# and make staticx pack them as well.
+# using scl here to get the proper LD_LIBRARY_PATH set
+RUN source scl_source enable devtoolset-8 llvm-toolset-7 && libs=$(./scripts/list_needed_libs.sh) && staticx $libs -l /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so dist/gprofiler dist/gprofiler
```

`libs=$(./scripts/list_needed_libs.sh)` already bring the `libgcc_s.so`, so there is no need to pass it again.
I've checked that the files are identical by running `cmp /lib64/libgcc_s.so.1 /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so`
`scripts/list_needed_libs.sh` do a great work :)

```suggestion
RUN source scl_source enable devtoolset-8 llvm-toolset-7 && libs=$(./scripts/list_needed_libs.sh) && staticx $libs dist/gprofiler dist/gprofiler
```

---

This is my output:
```
#29 0.353 found needed lib: /lib64/libpthread.so.0
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangFrontend.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangSerialization.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangDriver.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangParse.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangSema.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangCodeGen.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangAnalysis.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangRewrite.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangEdit.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangAST.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangLex.so.5
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/libclangBasic.so.5
#29 0.353 found needed lib: /lib64/librt.so.1
#29 0.353 found needed lib: /lib64/libdl.so.2
#29 0.353 found needed lib: /lib64/libtinfo.so.5
#29 0.353 found needed lib: /lib64/libz.so.1
#29 0.353 found needed lib: /lib64/libstdc++.so.6
#29 0.353 found needed lib: /lib64/libm.so.6
#29 0.353 found needed lib: /lib64/libgcc_s.so.1
#29 0.353 found needed lib: /lib64/libc.so.6
#29 0.353 found needed lib: /opt/rh/llvm-toolset-7/root/usr/lib64/../lib64/libLLVM-5.0.so
#29 0.353 found needed lib: /lib64/libffi.so.6
#29 0.353 found needed lib: /lib64/libc.so.6
#29 0.353 found needed lib: /lib64/libdl.so.2
#29 0.353 found needed lib: /lib64/libpthread.so.0
#29 0.353 found needed lib: /lib64/librt.so.1
#29 0.353 found needed lib: /lib64/libstdc++.so.6
#29 0.353 found needed lib: /lib64/libm.so.6
#29 0.353 found needed lib: /lib64/libgcc_s.so.1
#29 0.353 found needed lib: /lib64/libc.so.6
```

Doesn't seem like the same file - running on a clean `centos:7`:
```
[root@d922e9e023b8 /]# ls -l /lib64/libgcc_s.so.1
lrwxrwxrwx 1 root root 28 Nov 13 01:53 /lib64/libgcc_s.so.1 -> libgcc_s-4.8.5-20150702.so.1
[root@d922e9e023b8 /]# ls -l /lib64/libgcc_s.so.1^C
[root@d922e9e023b8 /]# ls /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so
ls: cannot access /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so: No such file or directory
```

I can try removing the manual `libgcc` reference and see if it works. Remind me, what exactly wasn't working for which we added this libgcc?

---

I think it was a steps to support `alpine` and `ubuntu 14.04`.
Anyway i compared the files in `pyi.Dockerfile` based container.
```bash
[root@4c04088634dd app]# cmp /lib64/libgcc_s.so.1 /usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so 
[root@4c04088634dd app]# 
```
So I think you can safely delete it.

---

Hmm. Weird tho because one is `libgcc_s-4.8.5-20150702.so.1` and the other is `/usr/lib/gcc/x86_64-redhat-linux/4.8.2/libgcc_s.so`, *4.8.2*...

But alright - will remove and see if CI passes on alpine :)

---

Seems to pass :)

---

Hmm no alpine tests. But those that we have seem to pass..